java-checks/src/main/java/org/sonar/java/checks/design/ClassImportCouplingCheck.java

```diff
@@ -0,0 +1,152 @@
+public class ClassImportCouplingCheck extends AbstractCouplingChecker {
+
+  // TODO change to 20, 3 is for testing purposes
+  private static final int COUPLING_THRESHOLD = 3;
+  @RuleProperty(
+    key = "couplingThreshold",
+    description = "Maximum number of classes a single class is allowed to depend upon",
+    defaultValue = "" + COUPLING_THRESHOLD)
+  public int couplingThreshold = COUPLING_THRESHOLD;
+  private String packageName;
+
+  @Override
+  public void visitClass(ClassTree tree) {
+    // if class is utility class -> don't report
+    if (UtilClassUtils.isUtilityClass(tree) || UtilClassUtils.isPrivateInnerClass(tree)) {
+      return;
+    }
+
+    if (tree.is(Tree.Kind.CLASS) && tree.simpleName() != null) {
+      nesting.push(types);
+      types = new HashSet<>();
+    }
+    CompilationUnitTree compilationUnitTree = (CompilationUnitTree) ExpressionUtils.getParentOfType(tree, Tree.Kind.COMPILATION_UNIT);
+    packageName = JavaTree.PackageDeclarationTreeImpl.packageNameAsString(compilationUnitTree.packageDeclaration());
+
+    Set<String> imports = compilationUnitTree.imports().stream()
+      .map(ImportTree.class::cast)
+      .map(ImportTree::qualifiedIdentifier)
+      .map(ClassImportCouplingCheck::helper)
+      .collect(Collectors.toSet());
+
+    Set<String> filteredImports = imports.stream()
+      .filter(i -> !i.startsWith("java."))
+      .collect(Collectors.toSet());
+
+    checkTypes(tree.superClass());
+    checkTypes(tree.superInterfaces());
+    super.visitClass(tree);
+
+    if (tree.is(Tree.Kind.CLASS) && tree.simpleName() != null) {
+      filteredImports.addAll(types);
+      int size = filteredImports.size();
+      if (size > couplingThreshold) {
+        context.reportIssue(this, tree, "Split this class into smaller and more specialized ones to reduce its dependencies on other classes from " +
+          size + " to the maximum authorized " + couplingThreshold + " or less.");
+      }
+      types = nesting.pop();
+    }
+  }
+
+  @Override
+  public void checkTypes(@Nullable Tree type) {
+    if (type == null || types == null) {
+      return;
+    }
+    if (type.is(Tree.Kind.IDENTIFIER)) {
+      IdentifierTreeImpl identifierTree = (IdentifierTreeImpl) type;
+      IPackageBinding packageBinding = identifierTree.typeBinding.getPackage();
+      addToTypes(identifierTree, packageBinding);
+    } else if (type.is(Tree.Kind.MEMBER_SELECT)) {
+      Deque<String> fullyQualifiedNameComponents = new ArrayDeque<>();
+      ExpressionTree expr = (ExpressionTree) type;
+      while (expr.is(Tree.Kind.MEMBER_SELECT)) {
+        MemberSelectExpressionTree mse = (MemberSelectExpressionTree) expr;
+        IdentifierTreeImpl identifierTree = (IdentifierTreeImpl) mse.identifier();
+        String name = identifierTree.binding.getName();
+        if (name.contains(packageName)) {
+          fullyQualifiedNameComponents.push(mse.identifier().name());
+        }
+        expr = mse.expression();
+      }
```

Comment:
Just to be sure I understand this block: If the class we are visiting depends upon a type from a subpackage, we only add the name of the identifier?
```java
package com.acme;

class MyType {
  void doNothing() {
    com.acme.sub.package.OtherType unused = null;
  }  
}
```
In the case above, is `OtherType` or ` com.acme.sub.package.OtherType` added to `types`?